File: src/quotes/selector.ts
import type { Quote, QuoteContext, SessionOutcome } from '../types.js';
import { QUOTES } from './corpus.js';

const AVOIDANCE_PATTERNS = [
  /\bi don'?t (?:want to|care|think)\b/i,
  /\bcan'?t you just\b/i,
  /\bjust tell me\b/i,
  /\bwhat'?s the (?:fix|answer|solution)\b/i,
  /\bi'?m not sure (?:why|how)\b/i,
  /\bthat'?s not (?:relevant|important)\b/i,
  /\bskip\b/i,
  /\bwhatever\b/i,
];

const OVERWHELM_PATTERNS = [
  /\bi don'?t know\b/i,
  /\bi'?m (?:lost|confused|stuck|overwhelmed)\b/i,
  /\bno idea\b/i,
  /\bthis is too\b/i,
  /\bi can'?t figure\b/i,
  /\bhelp\b/i,
  /\bi'?m not getting\b/i,
  /\bgive up\b/i,
];

const STRATEGY_PATTERNS = [
  /\bwhere (?:do i|should i) start\b/i,
  /\bso many (?:things|options|possibilities)\b/i,
  /\bmaybe (?:it'?s|i should)\b/i,
  /\bor maybe\b/i,
  /\bcould be (?:this|that|anything)\b/i,
  /\bnot sure which\b/i,
];

export function detectContext(response: string): QuoteContext | null {
  for (const pattern of AVOIDANCE_PATTERNS) {
    if (pattern.test(response)) return 'avoidance';
  }
  for (const pattern of OVERWHELM_PATTERNS) {
    if (pattern.test(response)) return 'overwhelm';
  }
  for (const pattern of STRATEGY_PATTERNS) {
    if (pattern.test(response)) return 'strategy';
  }
  return null;
}

export function selectQuote(response: string): Quote | null {
  const context = detectContext(response);
  if (!context) return null;

  const matching = QUOTES.filter((q) => q.context === context);
  if (matching.length === 0) return null;

  return matching[Math.floor(Math.random() * matching.length)];
}

export function selectClosingQuote(outcome: SessionOutcome): Quote {
  const context: QuoteContext = outcome === 'solved' ? 'victory' : 'perseverance';
  const matching = QUOTES.filter((q) => q.context === context);
  return matching[Math.floor(Math.random() * matching.length)];
}

File: nietzschian_debugger/quotes/selector.py
"""Quote selection based on developer response context."""

from __future__ import annotations

import random
import re
from typing import Optional

from ..types import Quote, QuoteContext, SessionOutcome
from .corpus import QUOTES

AVOIDANCE_PATTERNS = [
    re.compile(r"\bi don'?t (?:want to|care|think)\b", re.IGNORECASE),
    re.compile(r"\bcan'?t you just\b", re.IGNORECASE),
    re.compile(r"\bjust tell me\b", re.IGNORECASE),
    re.compile(r"\bwhat'?s the (?:fix|answer|solution)\b", re.IGNORECASE),
    re.compile(r"\bi'?m not sure (?:why|how)\b", re.IGNORECASE),
    re.compile(r"\bthat'?s not (?:relevant|important)\b", re.IGNORECASE),
    re.compile(r"\bskip\b", re.IGNORECASE),
    re.compile(r"\bwhatever\b", re.IGNORECASE),
]

OVERWHELM_PATTERNS = [
    re.compile(r"\bi don'?t know\b", re.IGNORECASE),
    re.compile(r"\bi'?m (?:lost|confused|stuck|overwhelmed)\b", re.IGNORECASE),
    re.compile(r"\bno idea\b", re.IGNORECASE),
    re.compile(r"\bthis is too\b", re.IGNORECASE),
    re.compile(r"\bi can'?t figure\b", re.IGNORECASE),
    re.compile(r"\bhelp\b", re.IGNORECASE),
    re.compile(r"\bi'?m not getting\b", re.IGNORECASE),
    re.compile(r"\bgive up\b", re.IGNORECASE),
]

STRATEGY_PATTERNS = [
    re.compile(r"\bwhere (?:do i|should i) start\b", re.IGNORECASE),
    re.compile(r"\bso many (?:things|options|possibilities)\b", re.IGNORECASE),
    re.compile(r"\bmaybe (?:it'?s|i should)\b", re.IGNORECASE),
    re.compile(r"\bor maybe\b", re.IGNORECASE),
    re.compile(r"\bcould be (?:this|that|anything)\b", re.IGNORECASE),
    re.compile(r"\bnot sure which\b", re.IGNORECASE),
]


def detect_context(response: str) -> Optional[QuoteContext]:
    """Detect the emotional context of a developer response."""
    for pattern in AVOIDANCE_PATTERNS:
        if pattern.search(response):
            return "avoidance"
    for pattern in OVERWHELM_PATTERNS:
        if pattern.search(response):
            return "overwhelm"
    for pattern in STRATEGY_PATTERNS:
        if pattern.search(response):
            return "strategy"
    return None


def select_quote(response: str) -> Optional[Quote]:
    """Select a contextually appropriate quote based on the developer response."""
    context = detect_context(response)
    if context is None:
        return None

    matching = [q for q in QUOTES if q.context == context]
    if not matching:
        return None

    return random.choice(matching)


def select_closing_quote(outcome: SessionOutcome) -> Quote:
    """Select a closing quote based on session outcome."""
    context: QuoteContext = "victory" if outcome == "solved" else "perseverance"
    matching = [q for q in QUOTES if q.context == context]
    return random.choice(matching)
